src/passes/Inlining.cpp

```diff
@@ -81,10 +83,13 @@ struct FunctionInfo {
     }
     // more than one use, so we can't eliminate it after inlining,
     // so only worth it if we really care about speed and don't care
-    // about size, and if it's lightweight so a good candidate for
-    // speeding us up.
+    // about size, and if it's lightweight (hasn't loops of calls)
+    // so a good candidate for speeding us up.
+    if (hasCalls) {
+      return false;
+    }
     return options.optimizeLevel >= 3 && options.shrinkLevel == 0 &&
-           (lightweight || options.inlining.allowHeavyweight);
+           (!hasLoops || options.inlining.allowHeavyweight);
```

Comment:
> which means it probably should have been named "allowFunctionsWithLoops".

Also thinking about it.

> Also the comment above should be changed to explain why this is - atm I don't understand why calls are handled differently from loops. Before it was simple, "lightweight" meant it can't do any heavy computation because without loops and without calls it just runs code once and exits (while a "heavyweight" function before could have done an arbitrary amount of work).

I'll improve comment. Main goal of this changes is exclude only function with calls which potentially could cause to problems with recursive calls. So lightweight now splitted to `hasLoops` and `hasCalls` and only if `hasLoops` + `allowHeavyweight` will be inlined. Functions which contain any call never inlined that's why I made early return if this happened